src/app/MessageDef.cpp

```diff
@@ -0,0 +1,2725 @@
+    } while (0)
+
+static void PrettyPrintDM(bool aIsNewLine, const char * aFmt, ...)
+{
+    va_list args;
+    size_t ret;
+    size_t sizeLeft;
+
+    va_start(args, aFmt);
+
+    if (aIsNewLine)
+    {
+        if (gCurLineBufferSize)
+        {
+            // Don't need to explicitly NULL-terminate the string because
+            // snprintf takes care of that.
+            ChipLogDetail(DataManagement, "%s", gLineBuffer);
+            gCurLineBufferSize = 0;
+        }
+
+        for (uint32_t i = 0; i < gPrettyPrintingDepthLevel; i++)
+        {
+            if (sizeof(gLineBuffer) > gCurLineBufferSize)
+            {
+                sizeLeft = sizeof(gLineBuffer) - gCurLineBufferSize;
+                ret      = (size_t)(snprintf(gLineBuffer + gCurLineBufferSize, sizeLeft, "\t"));
+                if (ret > 0)
+                {
+                    gCurLineBufferSize += MIN(ret, sizeLeft);
+                }
+            }
+        }
+    }
+
+    if (sizeof(gLineBuffer) > gCurLineBufferSize)
+    {
+        sizeLeft = sizeof(gLineBuffer) - gCurLineBufferSize;
+        ret      = (size_t)(vsnprintf(gLineBuffer + gCurLineBufferSize, sizeLeft, aFmt, args));
+        if (ret > 0)
+        {
+            gCurLineBufferSize += MIN(ret, sizeLeft);
+        }
+    }
+
+    va_end(args);
+}
+#else // CHIP_DETAIL_LOGGING
+#define PRETTY_PRINT_CHECKPOINT()
+#define PRETTY_PRINT(fmt, ...)
+#define PRETTY_PRINT_SAMELINE(fmt, ...)
+#define PRETTY_PRINT_INCDEPTH()
+#define PRETTY_PRINT_DECDEPTH()
+#endif // CHIP_DETAIL_LOGGING
+
+CHIP_ERROR LookForElementWithTag(const chip::TLV::TLVReader & aSrcReader, const uint64_t aTagInApiForm,
+                                  chip::TLV::TLVReader * apDstReader)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    // make a copy of the Path reader
+    chip::TLV::TLVReader reader;
+    reader.Init(aSrcReader);
+
+    while (CHIP_NO_ERROR == (err = reader.Next()))
+    {
+        // Documentation says the result of GetType must be verified before calling GetTag
+        VerifyOrExit(chip::TLV::kTLVType_NotSpecified != reader.GetType(), err = CHIP_ERROR_INVALID_TLV_ELEMENT);
+
+        if (aTagInApiForm == reader.GetTag())
+        {
+            apDstReader->Init(reader);
+            break;
+        }
+    }
+
+exit:
+    ChipLogIfFalse((CHIP_NO_ERROR == err) || (CHIP_END_OF_TLV == err));
+
+    return err;
+}
+
+ParserBase::ParserBase() { }
+
+CHIP_ERROR ParserBase::GetReaderOnTag(const uint64_t aTagToFind, chip::TLV::TLVReader * const apReader) const
+{
+    return LookForElementWithTag(mReader, aTagToFind, apReader);
+}
+
+template <typename T>
+CHIP_ERROR ParserBase::GetUnsignedInteger(const uint8_t aContextTag, T * const apLValue) const
+{
+    return GetSimpleValue(aContextTag, chip::TLV::kTLVType_UnsignedInteger, apLValue);
+}
+
+template <typename T>
+CHIP_ERROR ParserBase::GetSimpleValue(const uint8_t aContextTag, const chip::TLV::TLVType aTLVType, T * const apLValue) const
```

Comment:
can you think of a way to bind the aTLVType to the template type T?

Reply:
good idea, thanks for suggestion, would follow-up in this ticket, https://github.com/project-chip/connectedhomeip/issues/3646